lib/matchers/indexes.rb

```diff
@@ -0,0 +1,84 @@
+module Mongoid
+  module Matchers
+    class HaveIndexForMongoid3 # :nodoc:
+      def initialize(index_fields)
+        @index_fields = index_fields.symbolize_keys!
+      end
+
+      def with_options(options = { })
+        @options = options
+        self
+      end
+
+      def matches?(klass)
+        @klass  = klass.is_a?(Class) ? klass : klass.class
+        @errors = []
+
+        if @klass.respond_to?(:index_options)
```

Comment:
Since this code is quite different for different versions, break this up into v3/have_index_for.rb, v4/have_index_for.rb, etc. If the index is backward compatible, eg. index for v5 works for v6, just include that one as is.

Reply:
You might want to inherit common code, so `Mongoid::Matchers::HaveIndexFor < Mongoid::Matchers::HaveIndexFor::Base` or something like that.